Add support for Package URLs

alilleybrinker opened the pull request:
(Depends on https://github.com/CVEProject/cve-schema/pull/391; will need to be rebased on `develop` if/when that is merged, before this can be merged)

Introduce support for Package URLs in the CVE record format.

For more background on the topic of software identification, review CISA's ["Software Identification Ecosystem Option Analysis"](https://www.cisa.gov/resources-tools/resources/software-identification-ecosystem-option-analysis) paper.

> [!WARNING]
> __When reviewing, focus on the [last commit](https://github.com/CVEProject/cve-schema/pull/397/commits).__ This is a "Stacked PR," on top of #391, but GitHub shows both the commit making the `cpeApplicability` structure generic (from #391) and the commit adding purls (the part we care about here), when you click "Files changed" at the top of the PR, making review harder.

alilleybrinker committed:
feat: Make "cpeApplicability" structure generic.

This "renames" (not actually a rename, see below) the existing
"cpeApplicability" structure and its children from CPE-specific names
to generic names. For example, "cpeApplicability" becomes "applicability."

This is intended to permit future record format updates to add support for
additional kinds of software identifiers. This change itself does not add
any new kinds of software identifiers.

The prior "cpeApplicability" structure remains entirely supported, though
CNAs and any future ADPs enriching with software ID information should be
encouraged to use the more expressive new "applicability" structure instead,
and use of both at the same time should be treated as an error to avoid
ambiguity.

Signed-off-by: Andrew Lilley Brinker <alilleybrinker@gmail.com>

alilleybrinker committed:
feat: Add support for Package URLs.

Signed-off-by: Andrew Lilley Brinker <abrinker@mitre.org>

pombredanne commented:
@alilleybrinker re:
> Introduce support for OmniBOR Artifact IDs in the CVE record format.
You meant PURL here, right?

pombredanne commented:
Also questions originally posted by @alilleybrinker in https://github.com/CVEProject/cve-schema/issues/391#issuecomment-2722668371_
            

> 
> Feedback from QWG meeting:
> 
> - Constrain the purl schema to explicitly disallow version information.
>   - Would this demand vendoring the purl spec?

What do you mean?

> - How to handle the purl `generic` type?

There is nothing super special about it. What's at issue?

>   - Also a complexity with the `go` type, where the purl spec doesn't match Go's expressiveness (purl is case-insensitive while Go is case-sensitive)

This is a bug in the spec. It will be fixed.

> - How does this relate to the `affected` block, which also encodes version information?

Not sure if this is relevant, but we have a version range spec in PURL: The namespace and name must be lowercased.

pombredanne commented on schema/CVE_Record_Format.json:
```diff
@@ -500,6 +500,93 @@
+                {
+                    "anyOf": [
+                        { "required": ["cpeMatch"] },
+                        { "required": ["purlMatch"] }
```
Why not use `purl_match`?

alilleybrinker replied:
Ah, this appears to be a naming error. Will correct it.

alilleybrinker replied:
Ah, actually not a naming error. The field is called `purlMatch` to match the naming of `cpeMatch` present already. The internal name of the elements inside the `purlMatch` array is `purl_match` in the schema, a name which is not surfaced to users of the schema.

pombredanne commented on schema/CVE_Record_Format.json:
```diff
@@ -500,6 +500,93 @@
+                    "type": "boolean"
+                },
+                "criteria": {
+                    "description": "Placeholder until we find a formal purl schema",
```
Can you elaborate?

alilleybrinker replied:
Right now a purl is just defined as the URI type, meaning it's a URI but that we don't constrain it in this schema itself to be a purl. It would be up to the CVE Services application to apply additional validation to ensure URLs provided are package URLs, and not other types of URLs. Alternatively, we could try to incorporate a regular expression for purls, but I'm not aware of a readily-available option for this and am hesitant to write my own lest I get it wrong. The purl spec is sufficiently complicated that I consider writing its grammar as a regular expression non-trivial.

pombredanne commented on schema/CVE_Record_Format.json:
```diff
@@ -500,6 +500,93 @@
+            ]
+        },
+        "purl_match": {
+            "description": "purl match string or range",
```
What do you mean by purl_match?

alilleybrinker replied:
`purl_match` is the internal name of the object type as it exists in the schema, it is not a user-facing name, and is named thus to stylistically match the equivalent existing `cpe_match` structure.

pombredanne commented on schema/CVE_Record_Format.json:
```diff
@@ -500,6 +500,93 @@
+                "matchCriteriaId": {
+                    "$ref": "#/definitions/uuidType"
+                },
+                "versionStartExcluding": {
```
For version ranges, purl uses this https://github.com/package-url/purl-spec/blob/main/VERSION-RANGE-SPEC.rst

alilleybrinker replied:
Yes, our plan is to not use the `vers` spec, but instead to reuse the existing mechanism for expressing version constraints used for CPE. We expect this will both be easier to existing consumers of CVE data and will mitigate concerns around the `vers` spec's readiness for production use.

alilleybrinker commented:
> @alilleybrinker re:
> 
> > Introduce support for OmniBOR Artifact IDs in the CVE record format.
> > You meant PURL here, right?

Yes, this was an editing error made when I split the prior single PR into three. It's now fixed.

alilleybrinker commented:
> Also questions originally posted by @alilleybrinker in [#391 (comment)](https://github.com/CVEProject/cve-schema/pull/391#issuecomment-2722668371)_
> 
> > Feedback from QWG meeting:
> > 
> > * Constrain the purl schema to explicitly disallow version information.
> >   
> >   * Would this demand vendoring the purl spec?
> 
> What do you mean?
> 
> > * How to handle the purl `generic` type?
> 
> There is nothing super special about it. What's at issue?
> 
> > * Also a complexity with the `go` type, where the purl spec doesn't match Go's expressiveness (purl is case-insensitive while Go is case-sensitive)
> 
> This is a bug in the spec. It will be fixed.
> 
> > * How does this relate to the `affected` block, which also encodes version information?
> 
> Not sure if this is relevant, but we have a version range spec in PURL: The namespace and name must be lowercased.

1. Regarding vendoring purl, at issue is that the QWG doesn't want to generically incorporate "purl" without being rigorous about what version of purl is intended. Referencing `main` on the purl specification repository is not sufficient. The solutions would be to either maintain a copy of the purl spec from a specific commit hash to refer to a fixed version, or to wait until purl has been standardized by ECMA and refer to the ECMA-specified version.
2. On the `generic` type: members of the QWG raised questions about whether the `generic` type (and the `swid` type) would be usable for downstream consumers of purl data from CVE records, and whether it would make sense to disallow the `generic` or `swid` types so that the only usable types would be those referring to concrete platforms.
3. Regarding the `go` issue: if this is something to be fixed, that's great.

I'll add that the list above was intended just to capture issues raised during the meeting, and are not arguments from me, I'm just relaying them as best I can.

alilleybrinker reviewed (commented):
I do want to add a regular expression constraint to ensure that purls submitted are actually syntactically valid. We haven't yet reached a consensus on whether versions should be permissible in the purl itself, so that would need to be resolved to determine the exact contents of the regex.

The regex could also help protect against accidental submission of purls which are not URL encoded (and thus not valid).

---

purls should also probably be pulled out as their own type, both for current clarity and for possible reuse elsewhere in the schema where purls may become usable in the future.

alilleybrinker commented:
After discussion with the QWG, a Request for Discussion-style document is in the works to more completely address the design considerations in this proposal.

alilleybrinker commented:
I've gone through and responded to the open comments on this. The full motivation and reasoning behind the proposal is shared in the RFD for software identifiers (#407). Currently, the one remaining open issue tracked in the RFD is whether to extend the `affected` block instead of a generic variant of the `cpeApplicability` block. There's now an alternative PR (#409) implementing that.

@pombredanne, if you have a regular expression for validating Package URLs, I think we could consider using it instead of the more generic `uriType` the proposal uses now. Otherwise, we'd leave the full purl validation to be done by CVE Services instead of handling it in the cve-schema.

bj1064 reviewed (dismissed):
If not good don't merge use for public only

alilleybrinker commented:
Closing this, as the project has opted for an approach based on the `affected` array.

pombredanne commented:
@alilleybrinker can you elaborate a little?

alilleybrinker commented:
@pombredanne this PR added Package URLs by extending the `cpeApplicability` object in the schema. After discussion with the QWG on PR #407, we've opted to pursue an alternate design based on the `affected` array. The PR for that is #409.
